server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -545,8 +545,15 @@ public void updateShardState(
                         "Shard is marked as relocated, cannot safely move to state " + newRouting.state()
                     );
                 }
-            assert newRouting.active() == false || state == IndexShardState.STARTED || state == IndexShardState.CLOSED
-                : "routing is active, but local shard state isn't. routing: " + newRouting + ", local state: " + state;
+
+            if (newRouting.active() != false && state != IndexShardState.STARTED && state != IndexShardState.CLOSED) {
```

Comment:
I think this can only happen when the shard is a replica, since if it was or is now a primary, we would be in a new primary term, which would make the shard started message be skipped by the master. Can we assert that?

Reply:
Yes it looks like this can only affect replicas. I added a comment and assertion to that effect in [08cd9bb](https://github.com/elastic/elasticsearch/pull/87451/commits/08cd9bbb790b4ba50d67ab92cf1a795c6c0c0245).